From 00ce03c101b658e09e3ba54006ea5be280a56b4f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Amadeusz=20S=C5=82awi=C5=84ski?=
 <amadeuszx.slawinski@intel.com>
Date: Tue, 27 Nov 2018 16:25:54 +0100
Subject: [PATCH 239/292] ASoC: core: Refactor soc_cleanup_card_resources
 cleanup order
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Change order in which we unregister things to the one like in failpath
in snd_soc_instantiate_card(), otherwise we end up cleaning topology
after freeing card.

For this to work introduce snd_card_disconnect_sync() call at the
beginning, because we moved snd_card_free() to the end.

Change-Id: I5ddd51d4ab277e3d0ade12e500506d6b809bdf2f
Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@intel.com>
Reviewed-on:
---
 sound/soc/soc-core.c | 20 ++++++++++++--------
 1 file changed, 12 insertions(+), 8 deletions(-)

diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index a6760a37286d..8bf08799538c 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -2216,23 +2216,27 @@ static int soc_cleanup_card_resources(struct snd_soc_card *card)
 	for_each_card_rtds(card, rtd)
 		flush_delayed_work(&rtd->delayed_work);
 
-	/* free the ALSA card at first; this syncs with pending operations */
-	snd_card_free(card->snd_card);
-
-	/* remove and free each DAI */
-	soc_remove_dai_links(card);
-	soc_remove_pcm_runtimes(card);
+	/* sync pending file operations */
+	snd_card_disconnect_sync(card->snd_card);
 
 	/* remove auxiliary devices */
 	soc_remove_aux_devices(card);
 
-	snd_soc_dapm_free(&card->dapm);
-	soc_cleanup_card_debugfs(card);
+	/* remove and free each DAI */
+	soc_remove_dai_links(card);
 
 	/* remove the card */
 	if (card->remove)
 		card->remove(card);
 
+	snd_soc_dapm_free(&card->dapm);
+	soc_cleanup_card_debugfs(card);
+
+	/* free the ALSA card at last */
+	snd_card_free(card->snd_card);
+
+	soc_remove_pcm_runtimes(card);
+
 	return 0;
 }
 
-- 
2.21.0

